### PR TITLE
Fix silent fixed-point overflow (ESROGUE-735)

### PR DESCRIPTION
**Pull request by ruck314**
## Summary

- Fixed-point variables (`Fixed`/`UFixed`) now raise a `GeneralError` when set to a value outside the representable range, instead of silently clipping or storing an incorrect value
- Added `minValue()`/`maxValue()` to `Fixed` and `UFixed` Python model classes so the existing C++ range check in `Block::setFixed()` is activated
- Replaced the silent positive-edge-case clipping (`fPoint -= 1`) in C++ with an explicit overflow check that throws with the variable name, attempted value, and valid range

## Test plan

- [x] Existing 103 core tests pass
- [x] New `test_fixed_point_overflow_raises_error` covers: positive overflow, negative overflow, unsigned overflow, negative-into-unsigned, and boundary values
- [x] Verified with `Fixed(16, 14)` (the etaQ/etaI register scenario, range +-2.0) that overflow now produces a clear error message

### CODE REVIEW

File: src/rogue/interfaces/memory/Block.cpp [hunk 1]
```diff
@@ -2011,11 +2011,20 @@ void rim::Block::setFixed(const double& val, rim::Variable* var, int32_t index)
...
     int64_t fPoint = static_cast<int64_t>(round(val * pow(2, var->binPoint_)));
-    // Check for positive edge case
-    uint64_t mask = 1 << (var->valueBits_ - 1);
-    if (val > 0 && ((fPoint & mask) != 0)) {
-        fPoint -= 1;
-    }
+
+    // Compute representable range in integer domain
+    int64_t maxInt = (1LL << (var->valueBits_ - 1)) - 1;
+    int64_t minInt = -(1LL << (var->valueBits_ - 1));
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp]: Valid — fixed in be91c1b5c. Changed `1LL` to `1ULL` in both computations and cast back to `int64_t`, so the shift is well-defined even for `valueBits_==64`.

File: src/rogue/interfaces/memory/Block.cpp [hunk 2]
```diff
@@ -2011,11 +2011,20 @@ void rim::Block::setFixed(const double& val, rim::Variable* var, int32_t index)
...
+
+    // Check for overflow (rounding may push value beyond representable range)
+    if (fPoint > maxInt || fPoint < minInt)
+        throw(rogue::GeneralError::create("Block::setFixed",
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp]: Valid — fixed in be91c1b5c. `UFixed.maxValue()` now returns `(2**(bitSize-1) - 1) / (2**binPoint)` to match the signed two's-complement range that `setFixed()`/`getFixed()` actually enforce. The test assertion for `UFixedVar.maximum` was updated from 4095.9375 to 2047.9375 accordingly.

File: python/pyrogue/_Model.py [hunk 3]
```diff
@@ -826,3 +834,11 @@ def __init__(self, bitSize: int, binPoint: int) -> None:
...
         self.name = f'UFixed_{self.bitSize}_{self.binPoint}'
         self.ndType = np.dtype(np.float64)
+
+    def minValue(self) -> float:
+        """Return the minimum representable unsigned fixed-point value (0)."""
+        return 0.0
+
+    def maxValue(self) -> float:
+        """Return the maximum representable unsigned fixed-point value."""
+        return (2**self.bitSize - 1) / (2**self.binPoint)
```
[the remark this answers is not in the document]
Comment (ruck314) [reply python/pyrogue/_Model.py:844]: Valid — applied exactly as suggested in be91c1b5c. `UFixed.maxValue()` now returns `(2**(bitSize-1) - 1) / (2**binPoint)` to match the signed range the C++ conversion path actually supports.

File: tests/core/test_model_variables.py [hunk 4]
```diff
@@ -157,3 +157,57 @@ def test_remote_variable_metadata_exposes_model_constraints():
...
+            assert False, f"Expected error for overflow, got {got}"
+        except Exception:
+            pass  # Expected
+
+        try:
+            root.Dev.UFixedVar.set(-1.0)
+            got = root.Dev.UFixedVar.get()
+            assert False, f"Expected error for negative unsigned, got {got}"
+        except Exception:
+            pass  # Expected
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/core/test_model_variables.py]: Valid — fixed in be91c1b5c. All `try/except Exception: pass` blocks replaced with `pytest.raises(Exception, match="Value range error")`, which is both stricter (verifies the error message) and consistent with the rest of the test suite.

File: tests/core/test_model_variables.py [hunk 5]
```diff
@@ -157,3 +157,57 @@ def test_remote_variable_metadata_exposes_model_constraints():
         assert root.Dev.FixedVar._base.signed is True
         assert root.Dev.UFixedVar._base.binPoint == 4
         assert root.Dev.UFixedVar._base.signed is False
+
+        # Fixed-point models expose representable min/max
+        assert root.Dev.FixedVar.minimum == -2048.0
+        assert root.Dev.FixedVar.maximum == 2047.9375
+        assert root.Dev.UFixedVar.minimum == 0.0
+        assert root.Dev.UFixedVar.maximum == 4095.9375
```
[the remark this answers is not in the document]
Comment (ruck314) [reply tests/core/test_model_variables.py:246]: Valid — fixed in be91c1b5c. Updated `UFixed.maxValue()` to return the signed max `(2**(bitSize-1) - 1) / (2**binPoint)`, and the test assertion is now `root.Dev.UFixedVar.maximum == 2047.9375`.

File: src/rogue/interfaces/memory/Block.cpp [hunk 6]
```diff
@@ -2011,11 +2011,20 @@ void rim::Block::setFixed(const double& val, rim::Variable* var, int32_t index)
...
     int64_t fPoint = static_cast<int64_t>(round(val * pow(2, var->binPoint_)));
-    // Check for positive edge case
-    uint64_t mask = 1 << (var->valueBits_ - 1);
-    if (val > 0 && ((fPoint & mask) != 0)) {
-        fPoint -= 1;
-    }
+
+    // Compute representable range in integer domain (use 1ULL to avoid UB for 64-bit widths)
+    int64_t maxInt = static_cast<int64_t>((1ULL << (var->valueBits_ - 1)) - 1);
+    int64_t minInt = -static_cast<int64_t>(1ULL << (var->valueBits_ - 1));
```
[the remark this answers is not in the document]
Comment (ruck314) [reply src/rogue/interfaces/memory/Block.cpp]: Valid — fixed in 21bb136b0. Changed `minInt = -static_cast<int64_t>(1ULL << (var->valueBits_ - 1))` to `minInt = -maxInt - 1`, which is well-defined for all widths including 64-bit (avoids the UB from negating INT64_MIN).

File: python/pyrogue/_Model.py [hunk 7]
```diff
@@ -826,3 +834,11 @@ def __init__(self, bitSize: int, binPoint: int) -> None:
...
         self.name = f'UFixed_{self.bitSize}_{self.binPoint}'
         self.ndType = np.dtype(np.float64)
+
+    def minValue(self) -> float:
+        """Return the minimum representable unsigned fixed-point value (0)."""
+        return 0.0
+
+    def maxValue(self) -> float:
+        """Return the maximum value supported by the current rim.Fixed conversion path."""
+        return (2**(self.bitSize - 1) - 1) / (2**self.binPoint)
```
Comment (tristpinsm) [inline python/pyrogue/_Model.py]: It seems `UFixed` is actually implemented as `Fixed` and has the corresponding range. This contradicts what is documented in `docs/src/pyrogue_tree/core/fixed_point_models.rst`

Comment (ruck314) [reply python/pyrogue/_Model.py]: This was a real bug, not just a docs contradiction.

Root cause: both `Fixed` and `UFixed` Python models set `modelId = rim.Fixed`, and there was only a single `rim::Fixed` constant in `Constants.h`. Every dispatch switch in `Variable.cpp` routed `rim::Fixed` to `Block::setFixed`/`getFixed`, which always use signed two's-complement math. No signedness was propagated into the C++ layer, so `UFixed` has effectively been an alias for `Fixed` the whole time — the range documented in `docs/src/pyrogue_tree/core/fixed_point_models.rst` never matched the actual behavior.

Fixed in 744230156 by giving `UFixed` its own path:

- New `rim::UFixed = 0x09` modelId (exported to Python).
- New `Block::setUFixed`/`getUFixed` that enforce `[0, 2^bitSize - 1] / 2^binPoint`, reject negative writes, and read back without sign extension.
- `setFixedPy`/`getFixedPy` dispatch on `modelId_` so the numpy array / list / scalar paths work for both signed and unsigned.
- `Variable.cpp` switches wire `rim::UFixed` to the new handlers (reusing the Py wrappers).
- `UFixed.modelId = rim.UFixed`, `maxValue = (2**bitSize - 1) / (2**binPoint)`.
- Tests: `test_fixed_point_overflow_raises_error` now covers the full unsigned range (writes of `3000.5` and `4095.9375` on `UFixed(16,4)`, rejects `4096.0` and `-1.0`). `test_model.py` updated to expect the new modelId.

File: src/rogue/interfaces/memory/Variable.cpp [hunk 8]
```diff
@@ -755,6 +765,7 @@ std::string rim::Variable::getDumpValue(bool read) {
                 break;
 
             case rim::Fixed:
+            case rim::UFixed:
```
Comment (tristpinsm) [inline src/rogue/interfaces/memory/Variable.cpp:771]: this should probably be a separate case which calls `getUFixed`

**Review by tristpinsm — COMMENTED** [review]
Looks great! One more comment that might need addressing

File: src/rogue/interfaces/memory/Variable.cpp (hunk 8 above)
Comment (ruck314) [reply src/rogue/interfaces/memory/Variable.cpp:771]: Good catch — split into explicit `Fixed` and `UFixed` cases in 78fb67ca6. The previous fall-through happened to work because `getFixed_` is initialized to `&Block::getUFixed` for UFixed variables (Variable.cpp:341), but it was misleading to read.

**Review by tristpinsm — APPROVED** [review]

File: python/pyrogue/_Model.py (hunk 7 above)
Comment (ruck314) [reply python/pyrogue/_Model.py]: Addressed. In 744230156 `UFixed` was given its own `rim.UFixed` modelId with dedicated `Block::setUFixed` / `Block::getUFixed` paths (rejecting negatives, no sign extension on read), so it now matches the `[0, 2^bitSize - 1] / 2^binPoint` range documented in `fixed_point_models.rst`. In 411ababe6 I also updated the "How Conversion Works" paragraph in that doc to describe the split `rim.Fixed` / `rim.UFixed` dispatch instead of the old single-modelId wording.
